### PR TITLE
Fix: FLUX.2 encode_prompt dtype and add bitsandbytes int8+bfloat16 warning

### LINKED ISSUE #13772: Bad image output for Flux.2-dev, rocm, quantization and separate prompt encoding sequence
### Describe the bug

Image generation using `black-forest-labs/Flux.2-dev` with diffusers using quantization at int8 and separate stages for prompt encoding and transformer inference results in bad random checkerboard image output. This same workflow works find with all other large models I have tried (QwenImage, z-image, Flux.1-dev, stable-diffusion-3-5-large).

System:
 OS: Fedora 
 Kernel: x86_64 Linux 7.0.8-100.fc43.x86_64
 CPU: Intel Xeon Silver 4114 @ 40x 3GHz [46.0°C]
 GPU: AMD Radeon Pro W7900 (radeonsi, navi31, LLVM 21.1.8, DRM 3.64, 7.0.8-100.fc43.x86_64)
 RAM: 321061MiB

Using docker images: `rocm/pytorch`
tags tested:
- latest (as of May 20, 2026)
- rocm7.2.2_ubuntu24.04_py3.12_pytorch_release_2.7.1

Model resulting in bad output:
- black-forest-labs/Flux.2-dev

The reproduction script included runs the prompt encoding and inference with  an int8 quantization, but explicitly separated by unloading everything in between. 

Output image:

Image

### Reproduction

```python
import gc

import diffusers
import torch
import transformers

# tested with these docker images (rocm/pytorch):
#   rocm/pytorch:rocm7.2.2_ubuntu24.04_py3.12_pytorch_release_2.7.1
#   rocm/pytorch:latest (as of 2026-05-20)
# Where latest was at pytorch version 2.8.0

# this seems to make no difference on the output or performance
# torch.backends.cuda.enable_mem_efficient_sdp(False)

model = "black-forest-labs/FLUX.2-dev"
outfile = "cool-cat.png"
prompt = "A cat with a banjo"

print("==== Phase 1: text encoder ====")
print("Loading text encoder (quantization config: llm_int8)...")
te_qconfig = transformers.BitsAndBytesConfig(
    load_in_8bit=True,
)
text_encoder = transformers.Mistral3ForConditionalGeneration.from_pretrained(
    model,
    subfolder="text_encoder",
    quantization_config=te_qconfig,
    tie_word_embeddings=False,
    torch_dtype=torch.bfloat16,
)

print("Building prompt-encoder pipeline (with quantization)...")
encoder_pipeline = diffusers.Flux2Pipeline.from_pretrained(
    model,
    text_encoder=text_encoder,
    transformer=None,
    vae=None,
    torch_dtype=torch.bfloat16,
)

encoder_pipeline.to("cuda")

print("Encoding prompt...")
with torch.no_grad():
    prompt_embeds, text_ids = encoder_pipeline.encode_prompt(prompt=prompt)

print("Unloading prompt-encoder pipeline...")
del encoder_pipeline
del text_encoder
gc.collect()
torch.cuda.empty_cache()

print("==== Phase 2: inference ====")
print("Loading transformer (quantization config: llm_int8)...")
tr_qconfig = diffusers.BitsAndBytesConfig(
    load_in_8bit=True,
)
transformer = diffusers.Flux2Transformer2DModel.from_pretrained(
    model,
    subfolder="transformer",
    quantization_config=tr_qconfig,
    torch_dtype=torch.bfloat16,
)

print("Building inference pipeline (with quantization)...")
pipeline = diffusers.Flux2Pipeline.from_pretrained(
    model,
    text_encoder=None,
    tokenizer=None,
    transformer=transformer,
    torch_dtype=torch.bfloat16,
)

pipeline = pipeline.to("cuda")

print("Running inference...")
result = pipeline(prompt_embeds=prompt_embeds)

print(f"Saving image to {outfile}...")
result.images[0].save(outfile)

print("Done.")
```

### Logs

```shell
# python test-flux2-int8.py
==== Phase 1: text encoder ====
Loading text encoder (quantization config: llm_int8)...
Loading weights: 100%|██████████████████████████████████████████████████████████████████████████████████████████████| 585/585 [04:38<00:00,  2.10it/s]
Building prompt-encoder pipeline (with quantization)...
Loading pipeline components...: 100%|███████████████████████████████████████████████████████████████████████████████████| 3/3 [00:02<00:00,  1.44it/s]
Encoding prompt...
[transformers] Kwargs passed to `processor.__call__` have to be in `processor_kwargs` dict, not in `**kwargs`
/opt/venv/lib/python3.12/site-packages/bitsandbytes/autograd/_functions.py:123: UserWarning: MatMul8bitLt: inputs will be cast from torch.bfloat16 to float16 during quantization
  warnings.warn(f"MatMul8bitLt: inputs will be cast from {A.dtype} to float16 during quantization")
Unloading prompt-encoder pipeline...
==== Phase 2: inference ====
Loading transformer (quantization config: llm_int8)...
Loading checkpoint shards: 100%|████████████████████████████████████████████████████████████████████████████████████████| 7/7 [06:17<00:00, 53.96s/it]
Building inference pipeline (with quantization)...
Loading pipeline components...: 100%|███████████████████████████████████████████████████████████████████████████████████| 3/3 [00:01<00:00,  1.51it/s]
Running inference...
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 50/50 [05:16<00:00,  6.33s/it]
Saving image to cool-cat.png...
Done.
```

### System Info

- 🤗 Diffusers version: 0.38.0
- Platform: Linux-7.0.8-100.fc43.x86_64-x86_64-with-glibc2.39
- Running on Google Colab?: No
- Python version: 3.12.3
- PyTorch version (GPU?): 2.8.0+rocm7.0.0.git64359f59 (True)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Huggingface_hub version: 1.15.0
- Transformers version: 5.8.1
- Accelerate version: 1.13.0
- PEFT version: 0.19.1
- Bitsandbytes version: 0.49.2
- Safetensors version: 0.8.0-rc.0
- xFormers version: not installed
- Accelerator: NA

System:
 OS: Fedora 
 Kernel: x86_64 Linux 7.0.8-100.fc43.x86_64
 Shell: zsh 5.9
 Resolution: 10240x2880
 DE: GNOME 49.7
 WM: Mutter
 WM Theme: Adwaita
 GTK Theme: Adwaita [GTK2/3]
 Icon Theme: Adwaita
 Font: Adwaita Sans 11
 CPU: Intel Xeon Silver 4114 @ 40x 3GHz [46.0°C]
 GPU: AMD Radeon Pro W7900 (radeonsi, navi31, LLVM 21.1.8, DRM 3.64, 7.0.8-100.fc43.x86_64)
 RAM: 321061MiB

### Who can help?

This is general use issue about regular inference with a base model.

@sayakpaul @DN6

### PR DESCRIPTION
**Pull request by CoderSATTY**
# What does this PR do?
Fixes #13772

Two related fixes for the `Flux2Pipeline`:

1. **Added a defensive warning for bitsandbytes 8-bit + bfloat16 quantization.** 
   This specific combination causes precision loss (bfloat16 is internally downcasted to float16 during MatMul) which accumulates and corrupts images in FLUX models. The warning alerts users at pipeline initialization and suggests using NF4 4-bit quantization instead.

2. **Fixed `encode_prompt()` for pre-computed embedding workflows.**
   - Skips prompt string formatting if embeddings are already provided.
   - Automatically casts pre-computed embeddings to the exact precision (`dtype`) and `device` expected by the pipeline. This prevents the silent precision mismatches that were causing corrupted/noisy outputs when passing embeddings between different pipeline instances.

### Testing
- Tested on NVIDIA A100-80GB via Modal. 
- Verified that the pipeline correctly warns when loaded with `BitsAndBytesConfig(load_in_8bit=True)` and `torch.bfloat16`.
- Verified that single-pipeline and two-phase workflows function perfectly and without warning when using `NF4` quantization.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [ ] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [x] Was this discussed/approved via a GitHub issue or the forum? Please add a link to it if that's the case: https://github.com/huggingface/diffusers/issues/13772
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

## Who can review?
@yiyixuxu @sayakpaul

### CODE REVIEW

**Review by CoderSATTY — COMMENTED** [review]
Hi @sayakpaul @yiyixuxu, just a gentle ping on this when you have a moment! I've kept the branch updated with main. Let me know if you need me to make any adjustments or add any specific tests. Happy to help get this across the finish line!

**Comment by CoderSATTY** [comment]
Friendly ping on this one!
 This fixes the encode_prompt dtype issue and adds a bitsandbytes int8+bfloat16 warning for Flux2Pipeline as reported in #13772. Happy to adjust the approach if needed.
cc @sayakpaul @yiyixuxu @DN6
